### PR TITLE
Refactor shell_batch.cpp

### PR DESCRIPTION
**Pull request by MeAreJeenius**

### CODE REVIEW

File: include/shell.h [hunk 1]
```diff
@@ -60,6 +60,10 @@ class BatchFile {
 	std::shared_ptr<BatchFile> prev = {}; // shared with Shell.bf
 	std::unique_ptr<CommandLine> cmd = {};
 	std::string filename{};
+
+private:
+	[[nodiscard]] std::string ExpandedBatchLine(std::string_view line);
```
Comment (kcgen) [inline include/shell.h:65]: Is the `line` sv being re-assigned inside the function?

If not, suggest making it read-only at the API level w/ `const` because this helps readers understand the intent.

File: include/shell.h [hunk 2]
```diff
@@ -60,6 +60,10 @@ class BatchFile {
...
+	[[nodiscard]] std::string GetLine();
```
Comment (kcgen) [inline include/shell.h:66]: Excellent use of `nodiscard` +1!

File: include/string_utils.h [hunk 3]
```diff
@@ -224,6 +224,7 @@ inline bool is_empty(const char *str) noexcept
 
 // case-insensitive comparisons
 bool ciequals(const char a, const char b);
+[[nodiscard]] bool ci_str_equals(std::string_view a, std::string_view b);
```
Comment (kcgen) [inline include/string_utils.h:227]: Does `iequals(a, b)` not handle two string_views?

File: src/shell/shell_batch.cpp [hunk 4]
```diff
@@ -19,275 +19,211 @@
...
+constexpr uint8_t ESC             = 27;
+constexpr uint8_t LINE_FEED       = '\n';
+constexpr uint8_t TAB             = '\t';
+constexpr uint8_t UNIT_SEPARATOR  = 31;
+
+[[nodiscard]] static bool FoundLabel(std::string_view line, std::string_view label);
+
+BatchFile::BatchFile(DOS_Shell* host, const char* const resolved_name,
+                     const char* const entered_name, const char* const cmd_line)
+        : echo(host->echo),
```
Comment (kcgen) [inline src/shell/shell_batch.cpp:40]: Is the `DOS_Shell` pointer changed or used as a temporary? (hopefully it can be `const` too)

File: src/shell/shell_batch.cpp [hunk 5]
```diff
@@ -19,275 +19,211 @@
...
-				std::string env;
-				if (shell->GetEnvStr(cmd_read,env)) {
-					const char* equals = strchr(env.c_str(),'=');
-					if (!equals) continue;
-					equals++;
-					append_cmd_write(equals);
-				}
-				cmd_read = first;
-			}
+
```
Comment (kcgen) [inline src/shell/shell_batch.cpp:122]: This function was doing a lot! So much easier to read and surely safer (as you mentioned the previous code could crash). :partying_face: 

Do you have a a bunch of batch files that you're testing with, to be sure that we're not regression or dropping functionality?

Also, there are some shell unit tests - maybe this batch file handling could be added, and you could add some positive / negative / errors etc type tests to exercise all the corner cases?

**Review by kcgen — COMMENTED** [review]
Very nice cleanup, @MeAreJeenius !

Mostly minor comments.

My main concern is that we don't have regressions; batch processing is so variable and nuanced, I'm not sure if there are some test batch files out there that we can feed through this to make sure all the corner cases are hit (and that we're not dropping functionality).

**Comment by kcgen** [comment]
@MeAreJeenius , are you on Linux or macOS? If so, can you run it through UBASAN:

``` shell
meson setup -Dbuildtype=debug -Db_sanitize=undefined,address build/ubasan
meson compile -C build/ubasan
```

Binary: `build/ubasan/dosbox`

Also, set this to catch all ASAN issues:

``` shell
export ASAN_OPTIONS=strict_string_checks=1:detect_stack_use_after_return=1:check_initialization_order=1:strict_init_order=1:detect_leaks=1
```

File: include/shell.h (hunk 1 above)
Comment (johnnovak) [reply include/shell.h:65]: +1

File: include/string_utils.h (hunk 3 above)
Comment (johnnovak) [reply include/string_utils.h:227]: +1

File: src/misc/string_utils.cpp [hunk 6]
```diff
@@ -209,6 +209,21 @@ bool ciequals(const char a, const char b)
 	return tolower(a) == tolower(b);
 }
 
+bool ci_str_equals(std::string_view a, std::string_view b)
+{
+	if (a.size() != b.size()) {
+		return false;
+	}
```
[the remark this answers is not in the document]
Comment (johnnovak) [reply src/misc/string_utils.cpp:216]: It's not just string viees, std::string behaves the way and is null-clean. For C++ string, zero bytes have no special significance as explained here:  https://stackoverflow.com/a/48255316

I think we should go with this and rather weed out all low-level C string hackery as we come across them.

File: src/shell/shell_batch.cpp [hunk 7]
```diff
@@ -19,275 +19,211 @@
...
-                     char const *const resolved_name,
-                     char const *const entered_name,
-                     char const *const cmd_line)
-        : file_handle(0),
-          location(0),
-          echo(host->echo),
+constexpr uint8_t ESC             = 27;
+constexpr uint8_t LINE_FEED       = '\n';
+constexpr uint8_t TAB             = '\t';
+constexpr uint8_t UNIT_SEPARATOR  = 31;
```
Comment (johnnovak) [inline src/shell/shell_batch.cpp:34]: While you're at it, you could convert the constant to UpperCamelCase as per our convention.

File: src/shell/shell_batch.cpp [hunk 8]
```diff
@@ -19,275 +19,211 @@
...
+	}
+
 	filename = totalname;
 	// Test if file is openable
-	if (!DOS_OpenFile(totalname,(DOS_NOT_INHERIT|OPEN_READ),&file_handle)) {
-		//TODO Come up with something better
-		E_Exit("SHELL:Can't open BatchFile %s",totalname);
+	if (!DOS_OpenFile(totalname, (DOS_NOT_INHERIT | OPEN_READ), &file_handle)) {
+		// TODO Come up with something better
+		E_Exit("SHELL:Can't open BatchFile %s", totalname);
```
Comment (johnnovak) [inline src/shell/shell_batch.cpp:56]: Should be a warning, not an error, and the message should be:

```
SHELL: Can't open batch file '%s'
```

You could do a cleanup pass of the log messages too if you like.

File: src/shell/shell_batch.cpp [hunk 9]
```diff
@@ -19,275 +19,211 @@
...
+{
+	std::string line = {};
+	while (line.empty()) {
+		line = GetLine();
+
+		if (line.empty()) {
+			return false;
+		}
+
+		auto colon_index = line.find_first_of(':');
```
Comment (johnnovak) [inline src/shell/shell_batch.cpp:79]: const auto

File: src/shell/shell_batch.cpp [hunk 10]
```diff
@@ -19,275 +19,211 @@
...
-	DOS_SeekFile(file_handle,&(this->location),DOS_SEEK_SET);
 
+	line = ExpandedBatchLine(line);
+	strncpy(lineout, line.c_str(), CMD_MAXLINE);
+	return true;
+}
+
+std::string BatchFile::GetLine()
+{
+	// Open the batchfile and seek to stored postion
```
Comment (johnnovak) [inline src/shell/shell_batch.cpp:102]: Batch file (two words)

File: src/shell/shell_batch.cpp [hunk 11]
```diff
@@ -19,275 +19,211 @@
...
 
 void BatchFile::Shift()
 {
 	cmd->Shift(1);
 }
+
+static bool FoundLabel(std::string_view line, std::string_view label)
+{
+	auto label_start = line.find_first_not_of("=\t :");
+	auto pfx         = line.substr(0, label_start);
```
[the remark this answers is not in the document]
Comment (johnnovak) [reply src/shell/shell_batch.cpp:214]: Many of these functions are good candidates for unit tests.

File: include/shell.h (hunk 1 above)
Comment (kcgen) [reply include/shell.h:65]: I'm OK with that, and I see both points.

As a caller of the function prototype, my IDE expands the prototype and I see it takes in a `std::string_view`: Ok, I can rest assured my string won't be modified - that's as much as I need to know.

As a co-maintainer of the function body, `const` on the  arguments (yes, even scalars) tells me how the author intends to use the variable, and how I need to track it in my short-term memory.  

Non-`const` (even on scalars) in the body args says to the reader _"the author will/may/can re-assign this arg any point in the body, like a temporary variable in the code, so be on the look out for that."_

`const` (even on scalars) in the body args says to the reader _"the author is saying (and proving) that the arg's value stays the same throughout the body, so dear reader, you don't need to keep track of this one."

So if you really don't want to const the prototypes: OK, I get it, and as a caller I don't need to know more.

But on the function bodies: yes, please const (even sclars), because it's we're talking about communicating intention to readers/reviewers/maintainers - and `const` does the job very well.

File: src/misc/string_utils.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (kcgen) [reply src/misc/string_utils.cpp]: Or worded another way - is it possible for two `string_view`s to have  different `.size()`s but have the same null-terminated string content? (I think it's possible; so maybe the size-compare can be dropped?).

File: src/shell/shell_batch.cpp (hunk 7 above)
Comment (kcgen) [reply src/shell/shell_batch.cpp]: For the readers, assigning magic values to nice `constexpr` names is what we need - and that's all I really see.

And for you overhauling this code, I think it's fine however you want to write out those magic literal values (hex, dec, escapes, etc..).  In general we try to match up magic literals with what's found in spec sheets, docs, etc. 

ASCII has all of them, so really all good.

File: src/shell/shell_batch.cpp (hunk 5 above)
Comment (kcgen) [reply src/shell/shell_batch.cpp:108]: Fully agree it doesn't make sense to battle with gmocking the whole-program and thread a needle into these functions, , @MeAreJeenius.

The flatter more stand-alone refactored BatchFile class sounds great!

**Comment by kcgen** [comment]
Very nice refactoring, @MeAreJeenius! I think we're clear to merge :rocket: 

(PS- MSYS2 on GitHub's side has broke FluidSynth's unit tests; this isn't due to any change on the repo's side, as it was previously running in the last main build here: https://github.com/dosbox-staging/dosbox-staging/actions/runs/4440895580)
